SharedRootBatchEventsSource.cc

```diff
@@ -0,0 +1,287 @@
+          eventsTree_->GetEntry(nextEntry_++);
+
+          auto start = std::chrono::high_resolution_clock::now();
+          //determine uncompressed size
+          const auto entriesInOffset = laneInfos_[iLane].dataProducts_.size()+1;
+          unsigned int summedSizes=0;
+          for(int index = 0; index < eventIDs_.size(); ++index) {
+            //the last entry in the offsets is the uncompressed size for that event
+            summedSizes += offsetsAndBuffer_.first[(index+1)*entriesInOffset-1];
+          }
+          uncompressedBuffer_ = pds::uncompressBuffer(this->compression_, offsetsAndBuffer_.second, summedSizes);
+          //std::cout <<"compressed buffer size "<<offsetsAndBuffer_.second.size() <<std::endl;
+          //std::cout <<"uncompressed buffer size "<<uncompressedBuffer_.size() <<std::endl;
+          offsetsAndBuffer_.second = std::vector<char>(); //free memory
+          laneInfos_[iLane].decompressTime_ += 
+            std::chrono::duration_cast<decltype(laneInfos_[iLane].decompressTime_)>(std::chrono::high_resolution_clock::now() - start);
+
+          cachedEventIndex_ = 0;
+        }
+        laneInfos_[iLane].eventID_ = eventIDs_[cachedEventIndex_];
+        
+        const auto entriesInOffset = laneInfos_[iLane].dataProducts_.size()+1;
+        const unsigned int indexIntoOffsets = cachedEventIndex_*entriesInOffset;
+        std::vector<uint32_t> offsets(offsetsAndBuffer_.first.begin()+indexIntoOffsets,
+                                      offsetsAndBuffer_.first.begin()+indexIntoOffsets+entriesInOffset);
+
+        unsigned int beginOffsetInBuffer = 0;
+        for(int index = 1; index <= cachedEventIndex_; ++index) {
+          beginOffsetInBuffer += offsetsAndBuffer_.first[index*entriesInOffset-1];
+        }
+        unsigned int endOffsetInBuffer = beginOffsetInBuffer + offsets.back();
+
+        std::vector<char> uBuffer(uncompressedBuffer_.begin()+beginOffsetInBuffer,
+                                  uncompressedBuffer_.begin()+endOffsetInBuffer);
+
+        ++cachedEventIndex_;
+        /*{
+           auto const& id = this->laneInfos_[iLane].eventID_;
+          std::cout <<"event entry "<<nextEntry_-1<<" cache index "<<cachedEventIndex_-1<<std::endl;
+          std::cout <<"ID "<<id.run<<" "<<id.lumi<<" "<<id.event<<std::endl;
+          std::cout <<"ubuffer size "<<uBuffer.size()<<std::endl;
+          std::cout <<"offset size "<<offsets.size()<<std::endl;
+          //for(auto b: buffer) {
+          //  std::cout <<"  "<<b<<std::endl;
+          //}
+          }*/
+
+        auto group = optTask.group();
+        group->run([this, offsets=std::move(offsets), uBuffer = std::move(uBuffer), task = optTask.releaseToTaskHolder(), iLane]() {
+            auto& laneInfo = this->laneInfos_[iLane];
+
+            auto start = std::chrono::high_resolution_clock::now();
+            //uBuffer.pop_back();
+            pds::deserializeDataProducts(&(*uBuffer.begin()), &(*uBuffer.end()), 
```

Comment:
Oops--dereferencing `std::vector::end()` is undefined behavior.

```suggestion
            pds::deserializeDataProducts(uBuffer.data(), uBuffer.data() + uBuffer.size(),
```

See https://en.cppreference.com/w/cpp/container/vector/data.

Reply:
Much better. Done.